handlers/xdelta_handler.c

```diff
@@ -61,6 +83,81 @@ main_alloc(void __attribute__ ((__unused__)) *opaque, size_t  items, usize_t  si
   return r;
 }
 
+/* XDelta is often request the same blocks of source data
+   due to the nature of how VCDIFF patch is working. This
+   results in substantial amount of fseek calls for the 
+   same source file segment. We are using simple LRU cache
+   to minimize disk io when possible. This speeds up patch
+   application on 1GB system partion more than twice
+*/
+static int
+init_block_cache(head_t* head) {
+    TAILQ_INIT(head);
+
+    for (int i = 0; i < DEFAULT_LRU_SIZE; i++) {
+        node_t* e = malloc(sizeof(node_t));
+        if (e == NULL) {
+            ERROR("cache malloc failed size %d", sizeof(node_t));
+            return -1;
+        }
+        memset(e, 0, sizeof(node_t));
+        if ((e->block = malloc(DEFAULT_BLOCK_SIZE)) == NULL) {
+            ERROR("cache malloc failed size %d", DEFAULT_BLOCK_SIZE);
+            return -1;
+        }
+        e->blockno = (xoff_t) -1;
+        e->size = DEFAULT_BLOCK_SIZE;
+        TAILQ_INSERT_TAIL(head, e, nodes);
+        e = NULL;
+    }
+    return 0;
+}
+
+/* look for block in cache or read from file on miss */
+static node_t*
+get_block(handler_data* handle, xd3_source* source) {
+    head_t* head = &handle->head;
+    xoff_t offset = source->blksize * source->getblkno;
+    node_t *e = NULL, *next = NULL;
+    
+    /* check cache */
+    TAILQ_FOREACH_SAFE(e, head, nodes, next) {
+        if (e->blockno == source->getblkno) {
+            TAILQ_PROMOTE(head, e, nodes);
+            return e;
+        }
+    }
+    /* cache miss, read from file */
```

Comment:
If I read this correctly we replace the block data in the least recently used node. This way we don't need to free and alloc

Reply:
Added!